### PR TITLE
Fetch Pull Request as part of Comment Event Conversion

### PR DESCRIPTION
**Pull request by Aayyush**
In this PR, we remove the idea of `maybePull` and `maybeHeadRepo` by fetching the pull request as part of comment event conversion. Previously, for some providers, we had to resolve the pull in the command runner which was provider specific, making the branching logic hard to follow. 

Fetching the PR as part of comment event conversion streamlines the workflow and is a good start to deprecating command runnner.

### CODE REVIEW

File: server/vcs/provider/github/converter/comment_event_test.go [hunk 1]
```diff
@@ -63,9 +97,110 @@ func TestCommentEvent_Convert(t *testing.T) {
...
+		BaseBranch: *pull.Base.Ref,
+		Author:     *pull.User.Login,
+		State:      models.OpenPullState,
+		BaseRepo:   modelRepo,
+	}
+
+	When(githubPullGetter.GetPullRequest(modelRepo, *comment.Issue.Number)).ThenReturn(&pull, nil)
+	commentEvent, err := subject.Convert(&comment)
+	Ok(t, err)
+	Equals(t, modelRepo, commentEvent.BaseRepo)
```
Comment (nishkrishnan) [inline server/vcs/provider/github/converter/comment_event_test.go:116]: for new code, i wanna stop using these helpers, we should just use:

`assert.NoError()` and `assert.Equal()`

File: server/vcs/provider/github/converter/comment_event_test.go [hunk 2]
```diff
@@ -63,9 +97,110 @@ func TestCommentEvent_Convert(t *testing.T) {
...
 	Equals(t, models.User{
 		Username: *comment.Comment.User.Login,
 	}, commentEvent.User)
 	Equals(t, *comment.Issue.Number, commentEvent.PullNum)
+	Equals(t, modelPull, *commentEvent.Pull)
+}
+
+func TestRunCommentCommand_GithubPullErrorf(t *testing.T) {
```
Comment (nishkrishnan) [inline server/vcs/provider/github/converter/comment_event_test.go:124]: why do these have f in the error?

File: server/events/command_runner.go [hunk 3]
```diff
@@ -308,44 +231,6 @@ func newCtx(ctx context.Context, repoFullName string, pullNum int) context.Conte
 	return context.WithValue(ctx, logging.PullNumKey, strconv.Itoa(pullNum))
 }
 
-func (c *DefaultCommandRunner) ensureValidRepoMetadata(
```
Comment (nishkrishnan) [inline server/events/command_runner.go:311]: beautiful

File: server/controllers/events/events_controller_test.go [hunk 4]
```diff
@@ -313,3 +318,12 @@ type noopCommentHandler struct{}
...
 }
+
+// func TestRunCommentCommand_GitlabMergeRequestErrorf(t *testing.T) {
+// 	t.Log("if getting the gitlab merge request fails an error should be logged")
+// 	vcsClient := setup(t)
+// 	ctx := context.Background()
+// 	When(gitlabGetter.GetMergeRequest(fixtures.GitlabRepo.FullName, fixtures.Pull.Num)).ThenReturn(nil, errors.New("err"))
+// 	ch.RunCommentCommand(ctx, fixtures.GitlabRepo, &fixtures.GitlabRepo, nil, fixtures.User, fixtures.Pull.Num, nil, time.Now())
+// 	vcsClient.VerifyWasCalledOnce().CreateComment(fixtures.GitlabRepo, fixtures.Pull.Num, "`Error: making merge request API call to GitLab: err`", "")
+// }
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/controllers/events/events_controller_test.go:329]: I modified this test for events controller instead.

File: server/events/command_runner.go [hunk 5]
```diff
@@ -180,12 +149,13 @@ func (c *DefaultCommandRunner) RunAutoplanCommand(ctx context.Context, baseRepo
 	autoPlanRunner.Run(cmdCtx, nil)
 }
 
+// TODO: Make pass actual pull and repo
```
[the remark this answers is not in the document]
Comment (Aayyush) [reply server/events/command_runner.go:152]: Meant to replace the `maybePull` and `maybeRepo` with `pull` and `repo` lol. I have removed the comment 😅

File: server/vcs/provider/github/converter/comment_event_test.go (hunk 2 above)
Comment (Aayyush) [reply server/vcs/provider/github/converter/comment_event_test.go]: No clue, I just copied it over lol! I have removed it from the test methods

File: server/events/command_runner.go [hunk 6]
```diff
@@ -45,7 +41,7 @@ type CommandRunner interface {
 	// RunCommentCommand is the first step after a command request has been parsed.
 	// It handles gathering additional information needed to execute the command
 	// and then calling the appropriate services to finish executing the command.
-	RunCommentCommand(ctx context.Context, baseRepo models.Repo, maybeHeadRepo *models.Repo, maybePull *models.PullRequest, user models.User, pullNum int, cmd *command.Comment, timestamp time.Time)
+	RunCommentCommand(ctx context.Context, baseRepo models.Repo, headRepo *models.Repo, pull *models.PullRequest, user models.User, pullNum int, cmd *command.Comment, timestamp time.Time)
```
Comment (nishkrishnan) [inline server/events/command_runner.go:44]: can we make these not pointers as well?

File: server/events/command_runner.go [hunk 7]
```diff
@@ -105,14 +77,11 @@ func buildCommentCommandRunner(
...
-	GithubPullGetter         GithubPullGetter
-	AzureDevopsPullGetter    AzureDevopsPullGetter
-	GitlabMergeRequestGetter GitlabMergeRequestGetter
-	DisableAutoplan          bool
-	EventParser              EventParsing
-	GlobalCfg                valid.GlobalCfg
-	StatsScope               tally.Scope
+	VCSClient       vcs.Client
+	DisableAutoplan bool
+	EventParser     EventParsing
```
Comment (nishkrishnan) [inline server/events/command_runner.go:82]: are we still using this here?

File: server/vcs/provider/github/converter/comment_event.go [hunk 8]
```diff
@@ -37,14 +44,96 @@ func (e CommentEventConverter) Convert(comment *github.IssueCommentEvent) (event
...
+		UpdatedAt:  updatedAt,
+	}
+	return
+}
+
+// ParseGithubRepo parses the response from the GitHub API endpoint that
+// returns a repo into the Atlantis model.
+func (c *CommentEventConverter) ParseGithubRepo(ghRepo *github.Repository) (models.Repo, error) {
+	return models.NewRepo(models.Github, ghRepo.GetFullName(), ghRepo.GetCloneURL(), c.RepoConverter.GithubUser, c.RepoConverter.GithubToken)
+}
```
Comment (nishkrishnan) [inline server/vcs/provider/github/converter/comment_event.go:139]: Why are you rewriting this? this already exists in the converters package https://github.com/lyft/atlantis/blob/6989c17c90b4b45e7be509bac71010ad254d8fdd/server/vcs/provider/github/converter/pull.go#L16-L17

File: server/vcs/provider/github/converter/comment_event_test.go [hunk 9]
```diff
@@ -49,23 +123,42 @@ func TestCommentEvent_Convert(t *testing.T) {
...
+
+	// act
+	subject := CommentEventConverter{
+		RepoConverter: repoConverter,
+		PullGetter:    githubPullGetter,
+	}
+	_, err := subject.Convert(&comment)
+
+	// assert
+	ErrContains(t, "getting pull from github", err)
```
Comment (nishkrishnan) [inline server/vcs/provider/github/converter/comment_event_test.go:144]: can you use assert here

File: server/vcs/provider/github/converter/comment_event_test.go [hunk 10]
```diff
@@ -49,23 +123,42 @@ func TestCommentEvent_Convert(t *testing.T) {
...
+
+	// act
+	subject := CommentEventConverter{
+		RepoConverter: repoConverter,
+		PullGetter:    githubPullGetter,
+	}
+	_, err := subject.Convert(&comment)
+
+	// assert
+	ErrContains(t, "converting pull request type", err)
```
Comment (nishkrishnan) [inline server/vcs/provider/github/converter/comment_event_test.go:163]: can you use assert here

File: server/events/command_runner.go (hunk 7 above)
Comment (Aayyush) [reply server/events/command_runner.go]: Nope, thanks for the catch :)

File: server/vcs/provider/github/converter/comment_event.go (hunk 8 above)
Comment (Aayyush) [reply server/vcs/provider/github/converter/comment_event.go]: Didn't realize it 😅! I removed it and reused the existing methods

**Review by nishkrishnan — APPROVED** [review]
